File: frontend/src/lib/academics.js
// Choice lists for a student's academic profile. COURSES is still a placeholder awaiting USTP–CDO's real programmes.

export const COURSES = [
  'BS Information Technology',
  'BS Computer Science',
  'BS Computer Engineering',
  'BS Civil Engineering',
  'BS Electrical Engineering',
  'BS Electronics Engineering',
  'BS Mechanical Engineering',
  'BS Chemical Engineering',
  'BS Architecture',
  'BS Accountancy',
  'BS Business Administration',
  'BS Food Technology',
  'BS Environmental Science',
  'BS Secondary Education',
];

// Mirrors the server (TrailSync/academics.py) and the two lines of the printed form; any number can be ticked at once.
export const ACADEMIC_STATUS_OPTIONS = [
  { value: 'Undergraduate Student', hint: 'Taking a bachelor’s degree now' },
  { value: 'Graduate/Masteral Student', hint: 'Taking a master’s or doctorate now' },
  { value: 'Alumnus — High School', hint: 'Finished high school at USTP' },
  { value: 'Alumnus — Undergraduate', hint: 'Finished a bachelor’s degree at USTP' },
  { value: 'Alumnus — Graduate/Masteral', hint: 'Finished a master’s or doctorate at USTP' },
];

export const isAlumnus = (statuses) => (statuses || []).some((s) => s.startsWith('Alumnus'));
// College alumni only: pre-2018 college records are in the archive, high school ones aren't.
export const hasCollegeAlumnus = (statuses) =>
  (statuses || []).some((s) => s === 'Alumnus — Undergraduate' || s === 'Alumnus — Graduate/Masteral');

// The ticked options in the form's order, for one line of text.
export function academicStatusLine(statuses) {
  const chosen = new Set(statuses || []);
  return ACADEMIC_STATUS_OPTIONS.filter((o) => chosen.has(o.value))
    .map((o) => o.value)
    .join(', ');
}

// "Last semester attended": the same format and plausibility rules the server applies, so mistakes show before saving.
export const SEMESTER_EXAMPLE = '2nd Semester, SY 2024-2025';
const SEMESTER_RE =
  /^\s*(?:(1st|first)\s*sem(?:ester)?|(2nd|second)\s*sem(?:ester)?|(summer))\s*,?\s*(?:s\.?\s*y\.?|school\s+year)\s*(\d{4})\s*[-–—]\s*(\d{4})\s*$/i;
const TERM_NAMES = ['1st Semester', '2nd Semester', 'Summer'];
const EARLIEST_SCHOOL_YEAR = 1950;

// Aug–Dec is the 1st semester, Jan–May the 2nd, Jun–Jul summer.
function currentTerm(today = new Date()) {
  const month = today.getMonth() + 1;
  if (month >= 8) return [today.getFullYear(), 0];
  if (month <= 5) return [today.getFullYear() - 1, 1];
  return [today.getFullYear() - 1, 2];
}

// { value } in the canonical spelling, or { error } in words a student can act on.
export function checkSemester(text) {
  if (!(text || '').trim()) return { error: 'Please enter the last semester you attended.' };
  const m = SEMESTER_RE.exec(text);
  if (!m) {
    return {
      error: `Please write it like “${SEMESTER_EXAMPLE}”: the semester (1st Semester, 2nd Semester or Summer), then SY and the school year.`,
    };
  }
  const start = Number(m[4]);
  const end = Number(m[5]);
  if (end !== start + 1) return { error: `A school year runs over two years in a row, like SY ${start}-${start + 1}.` };
  const term = m[1] ? 0 : m[2] ? 1 : 2;
  const [nowStart, nowTerm] = currentTerm();
  if (start > nowStart || (start === nowStart && term > nowTerm)) {
    return { error: 'That semester hasn’t happened yet. Enter the last one you actually attended.' };
  }
  if (start < EARLIEST_SCHOOL_YEAR) {
    return { error: `Please check the school year; it should be SY ${EARLIEST_SCHOOL_YEAR}-${EARLIEST_SCHOOL_YEAR + 1} or later.` };
  }
  return { value: `${TERM_NAMES[term]}, SY ${start}-${end}` };
}

File: BTrailSync/TrailSync/academics.py
"""Academic status options and the "last semester attended" format, shared by onboarding, the profile and requests."""
import re
from datetime import date

# Mirrors the two lines of FM-USTP-RGTR-09 ("Student (Undergrad/Graduate)" and "Alumnus (High School/Undergrad/Graduate)"),
# split so someone can be several at once, e.g. an undergraduate alumnus now in a master's programme.
UNDERGRADUATE_STUDENT = "Undergraduate Student"
GRADUATE_STUDENT = "Graduate/Masteral Student"
ALUMNUS_HIGH_SCHOOL = "Alumnus — High School"
ALUMNUS_UNDERGRADUATE = "Alumnus — Undergraduate"
ALUMNUS_GRADUATE = "Alumnus — Graduate/Masteral"

ACADEMIC_STATUS_OPTIONS = (
    UNDERGRADUATE_STUDENT,
    GRADUATE_STUDENT,
    ALUMNUS_HIGH_SCHOOL,
    ALUMNUS_UNDERGRADUATE,
    ALUMNUS_GRADUATE,
)
STUDENT_STATUSES = frozenset({UNDERGRADUATE_STUDENT, GRADUATE_STUDENT})
ALUMNUS_STATUSES = frozenset({ALUMNUS_HIGH_SCHOOL, ALUMNUS_UNDERGRADUATE, ALUMNUS_GRADUATE})
# College graduates only: the pre-2018 archive holds college records, not high school ones.
ARCHIVE_STATUSES = frozenset({ALUMNUS_UNDERGRADUATE, ALUMNUS_GRADUATE})


def is_student(statuses):
    return any(s in STUDENT_STATUSES for s in statuses or ())


def is_alumnus(statuses):
    return any(s in ALUMNUS_STATUSES for s in statuses or ())


def ordered_statuses(statuses):
    """The chosen options in the form's own order, without duplicates."""
    chosen = set(statuses or ())
    return [s for s in ACADEMIC_STATUS_OPTIONS if s in chosen]


# "1st Semester, SY 2024-2025", "2nd Semester, SY 2024-2025" or "Summer, SY 2024-2025". Case, spacing, "Sem" and "S.Y."
# are forgiven, then the value is stored in that one canonical spelling.
SEMESTER_EXAMPLE = "2nd Semester, SY 2024-2025"
_SEMESTER_RE = re.compile(
    r"^\s*(?:(?P<first>1st|first)\s*sem(?:ester)?|(?P<second>2nd|second)\s*sem(?:ester)?|(?P<summer>summer))"
    r"\s*,?\s*(?:s\.?\s*y\.?|school\s+year)\s*(?P<start>\d{4})\s*[-–—]\s*(?P<end>\d{4})\s*$",
    re.IGNORECASE,
)
TERM_NAMES = ("1st Semester", "2nd Semester", "Summer")
# A plausibility floor, not a founding date: nobody requesting records today last attended before this.
EARLIEST_SCHOOL_YEAR = 1950


def current_term(today=None):
    """(school-year start, term index) for today: Aug-Dec is the 1st semester, Jan-May the 2nd, Jun-Jul summer."""
    today = today or date.today()
    if today.month >= 8:
        return today.year, 0
    if today.month <= 5:
        return today.year - 1, 1
    return today.year - 1, 2


def normalize_semester(text, today=None):
    """The canonical spelling of a last-semester answer, or ValueError with a message a student can act on."""
    match = _SEMESTER_RE.match(text or "")
    if not match:
        raise ValueError(
            f"Please write it like “{SEMESTER_EXAMPLE}”: the semester (1st Semester, 2nd Semester or Summer), "
            "then SY and the school year."
        )
    start, end = int(match["start"]), int(match["end"])
    if end != start + 1:
        raise ValueError(f"A school year runs over two years in a row, like SY {start}-{start + 1}.")
    term = 0 if match["first"] else 1 if match["second"] else 2
    if (start, term) > current_term(today):
        raise ValueError("That semester hasn’t happened yet. Enter the last one you actually attended.")
    if start < EARLIEST_SCHOOL_YEAR:
        raise ValueError(f"Please check the school year; it should be SY {EARLIEST_SCHOOL_YEAR}-{EARLIEST_SCHOOL_YEAR + 1} or later.")
    return f"{TERM_NAMES[term]}, SY {start}-{end}"
